core/trino-main/src/main/java/io/trino/execution/scheduler/FullNodeCapableNodeAllocatorService.java

```diff
@@ -0,0 +1,595 @@
+
+                    detachedIterator.remove();
+                }
+                catch (RuntimeException e) {
+                    failures.put(pendingAcquire, e);
+                    detachedIterator.remove();
+                }
+            }
+
+            Set<InternalNode> nodes = ImmutableSet.copyOf(fullNodePendingAcquires.keySet());
+            for (InternalNode reservedNode : nodes) {
+                PendingAcquire pendingAcquire = fullNodePendingAcquires.get(reservedNode);
+                if (pendingAcquire.getFuture().isCancelled()) {
+                    // discard cancelled pendingAcquire with target node
+                    fullNodePendingAcquires.remove(reservedNode);
+                    verify(fullNodesByQueryId.remove(pendingAcquire.getQueryId(), reservedNode));
+                    continue;
+                }
+                try {
+                    Candidates currentCandidates = selectCandidates(pendingAcquire.getNodeRequirements(), pendingAcquire.getNodeSelector());
+                    if (currentCandidates.isEmpty()) {
+                        throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+                    }
+
+                    if (sharedAllocatedMemory.getOrDefault(reservedNode, 0L) > 0 || allocatedFullNodes.contains(reservedNode)) {
+                        // reserved node is still used - opportunistic check if maybe there is some other empty, not waited for node available
+                        Optional<InternalNode> opportunisticNode = currentCandidates.getCandidates().stream()
+                                .filter(node -> !fullNodePendingAcquires.containsKey(node))
+                                .filter(node -> !allocatedFullNodes.contains(node))
+                                .filter(node -> sharedAllocatedMemory.getOrDefault(node, 0L) == 0)
+                                .findFirst();
+
+                        if (opportunisticNode.isPresent()) {
+                            fullNodePendingAcquires.remove(reservedNode);
+                            verify(fullNodesByQueryId.remove(pendingAcquire.getQueryId(), reservedNode));
+                            allocatedFullNodes.add(opportunisticNode.get());
+                            verify(fullNodesByQueryId.put(pendingAcquire.getQueryId(), opportunisticNode.get()));
+                            assignedNodes.put(pendingAcquire, opportunisticNode.get());
+                        }
+                        continue;
+                    }
+
+                    if (!currentCandidates.getCandidates().contains(reservedNode)) {
+                        // current candidate is gone; move pendingAcquire to detached state
+                        detachedFullNodePendingAcquires.add(pendingAcquire);
+                        fullNodePendingAcquires.remove(reservedNode);
+                        // trigger one more round of processing immediatelly
+                        wakeupProcessPendingAcquires();
```

Comment:
`continue`?

Reply:
mhm - this actually can be tested - will add test for it.